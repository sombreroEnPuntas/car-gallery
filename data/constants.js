// @flow

export const CAR_SERVICE_URL = 'https://car-list-service.herokuapp.com/api'
