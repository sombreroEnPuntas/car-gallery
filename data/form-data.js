// @flow

// Data
import type { VehicleT } from './car-service'

// Utils
import { path, pathOr, compose, filter, propEq, always } from 'ramda'

// Components
import type { DropdownFieldT } from '../components/dropdown-field'

export const isVehicleKey: string => boolean = key =>
  [
    'bodyType',
    'engineCapacity',
    'enginePowerKW',
    'enginePowerPS',
    'fuelType',
  ].includes(key)

const getFieldPlaceholder = (key, state) => {
  if (key !== 'make' && !path(['formData', 'make', 'valid'], state)) {
    return { disabled: true, placeholder: 'Choose make!' }
  } else if (
    key !== 'make' &&
    key !== 'model' &&
    !path(['formData', 'model', 'valid'], state)
  ) {
    return { disabled: true, placeholder: 'Choose model!' }
  } else {
    return
  }
}

export const makeFieldHelper: (*, *) => string => DropdownFieldT = (
  state,
  handleUpdate
) => key => ({
  ...getFieldPlaceholder(key, state),
  handleUpdate: handleUpdate,
  name: key,
  // $FlowIgnore pathOr returns string when evaluated
  valid: pathOr(false, ['formData', key, 'valid'], state),
  // $FlowIgnore pathOr returns string when evaluated
  value: pathOr('', ['formData', key, 'value'], state),
  values: isVehicleKey(key)
    ? // $FlowIgnore compose returns array when evaluated
      getCarsFiltered(state)
    : // $FlowIgnore pathOr returns string when evaluated
      pathOr([], ['formData', key, 'values'], state),
})

export const getScore = (state: *) =>
  (path(['formData', 'make', 'valid'], state) ? 75 : 0) +
  (path(['formData', 'model', 'valid'], state) ? 75 : 0) +
  (path(['formData', 'bodyType', 'valid'], state) ? 25 : 0) +
  (path(['formData', 'engineCapacity', 'valid'], state) ? 25 : 0) +
  (path(['formData', 'enginePowerKW', 'valid'], state) ? 25 : 0) +
  (path(['formData', 'fuelType', 'valid'], state) ? 25 : 0)

export const getCarOptionsByKey: string => (
  Array<string>
) => Array<string> = key => data => {
  let options = []
  data.forEach(car => !options.includes(car[key]) && options.push(car[key]))
  return options
}

const getFilter = (key, value) =>
  !!value && value !== '' ? propEq(key, value) : always()

export const getCarsFiltered = (state: *) =>
  compose(
    // $FlowIgnore filter returns array when evaluated
    filter(
      getFilter(
        'bodyType',
        // $FlowIgnore pathOr returns string when evaluated
        pathOr('', ['formData', 'bodyType', 'value'], state)
      )
    ),
    // $FlowIgnore filter returns array when evaluated
    filter(
      getFilter(
        'engineCapacity',
        // $FlowIgnore pathOr returns string when evaluated
        pathOr('', ['formData', 'engineCapacity', 'value'], state)
      )
    ),
    // $FlowIgnore filter returns array when evaluated
    filter(
      getFilter(
        'enginePowerKW',
        // $FlowIgnore pathOr returns string when evaluated
        pathOr('', ['formData', 'enginePowerKW', 'value'], state)
      )
    ),
    // $FlowIgnore filter returns array when evaluated
    filter(
      getFilter(
        'enginePowerPS',
        // $FlowIgnore pathOr returns string when evaluated
        pathOr('', ['formData', 'enginePowerPS', 'value'], state)
      )
    ),
    // $FlowIgnore filter returns array when evaluated
    filter(
      getFilter(
        'fuelType',
        // $FlowIgnore pathOr returns string when evaluated
        pathOr('', ['formData', 'fuelType', 'value'], state)
      )
    )
    // $FlowIgnore filter returns array when evaluated
  )(pathOr([], ['formData', 'vehicle', 'values'], state))
