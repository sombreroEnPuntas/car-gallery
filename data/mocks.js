// @flow

export const makesList = ['ford', 'opel']
export const modelsList = ['explorer', 'fiesta']

// API
export const internalError = {
  error: `Wut?. SyntaxError`,
}

export const errorAPIResponse = { error: `418 I'm a teapot` }
