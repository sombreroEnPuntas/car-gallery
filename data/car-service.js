// @flow

// NextJS requires a polyfill
import fetch from 'isomorphic-unfetch'
import { CAR_SERVICE_URL } from './constants'

export type APIErrorResponseT = { error: boolean, message: string }

export type GetMakesResponseT = Array<string> | APIErrorResponseT
type GetMakesT = () => Promise<GetMakesResponseT>
export const getMakes: GetMakesT = async () => {
  let result

  try {
    const response = await fetch(`${CAR_SERVICE_URL}/makes`)
    const json = await response.json()

    // normalize output
    let makes = []
    json.forEach(make => makes.push(make.toLowerCase()))

    result = makes
  } catch (error) {
    result = { error: true, ...error }
  }

  return result
}

export type GetModelsResponseT = Array<string> | APIErrorResponseT
type GetModelsT = (
  make: string,
  makes: Array<string>
) => Promise<GetModelsResponseT>
export const getModels: GetModelsT = async (make, makes) => {
  let result

  try {
    // normalize input
    const param = make.toLowerCase()

    // validate input
    if (!makes.includes(param)) {
      throw {
        message: `Invalid input provided as parameter.`,
        type: 'bad-input',
      }
    }

    const response = await fetch(`${CAR_SERVICE_URL}/models?make=${make}`)
    const json = await response.json()

    // normalize output
    let models = []
    json.forEach(model => models.push(model.toLowerCase()))

    result = json
  } catch (error) {
    result = { error: true, ...error }
  }

  return result
}
