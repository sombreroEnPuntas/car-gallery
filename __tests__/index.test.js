// @flow
import React from 'react'
import { shallow } from 'enzyme'

// Data
import { makesList, modelsList } from '../data/mocks'

// Dependencies
import { getMakes, getModels } from '../data/car-service'

// Tested unit
import TestedComponent from '../pages/index.js'

// Utils
import { assocPath } from 'ramda'

// Mocks
jest.mock('../data/car-service')
const setGetMakesMock = (data, error) =>
  getMakes.mockImplementation(() =>
    Promise.resolve(error ? { error: `418 I'm a teapot` } : data)
  )
const setGetModelsMock = (data, error) =>
  getModels.mockImplementation(() =>
    Promise.resolve(error ? { error: `418 I'm a teapot` } : data)
  )
const consoleSpy = jest.spyOn(console, 'log')

const getProps = customProps => ({
  ...customProps,
})

describe('Index', () => {
  afterAll(() => {
    jest.restoreAllMocks() // clean .spyOn mocks
    jest.resetAllMocks() // clean .mock
  })

  it(`matches snapshot`, async done => {
    setGetMakesMock(makesList)
    setGetModelsMock(modelsList)

    const wrapper = await shallow(<TestedComponent {...getProps({})} />)

    process.nextTick(() => {
      expect(wrapper).toMatchSnapshot()

      getMakes.mockClear()
      getModels.mockClear()

      done()
    })
  })

  describe.each`
    displayName | error   | data                  | formData
    ${'make'}   | ${null} | ${makesList}          | ${{}}
    ${'make'}   | ${true} | ${`418 I'm a teapot`} | ${{}}
    ${'model'}  | ${null} | ${modelsList}         | ${{ make: { value: 'ford', valid: true, values: makesList } }}
    ${'model'}  | ${true} | ${`418 I'm a teapot`} | ${{ make: { value: 'ford', valid: true, values: makesList } }}
  `(`fetch $displayName`, ({ displayName, error, data, formData }) => {
    it(`fetches ${data}`, async done => {
      setGetMakesMock(makesList, error)
      setGetModelsMock(modelsList, error)

      const wrapper = await shallow(<TestedComponent {...getProps({})} />)

      wrapper.setState({
        error: null,
        formData,
        isLoading: true,
      })

      process.nextTick(() => {
        expect(wrapper.state()).toEqual({
          error: error ? data : null,
          formData: assocPath([displayName, 'values'], error ? null : data)(
            formData
          ),
          isLoading: false,
        })

        getMakes.mockClear()
        getModels.mockClear()

        done()
      })
    })

    it(`doesn't fetch if there's an error`, async done => {
      setGetMakesMock(makesList, error)
      setGetModelsMock(modelsList, error)

      const wrapper = await shallow(<TestedComponent {...getProps({})} />)

      wrapper.setState({
        error: `418 I'm a teapot`,
        formData,
        isLoading: true,
      })

      process.nextTick(() => {
        expect(wrapper.state()).toEqual({
          error: `418 I'm a teapot`,
          formData,
          isLoading: true,
        })

        expect(getMakes).toHaveBeenCalledTimes(1)
        expect(getModels).toHaveBeenCalledTimes(0)

        getMakes.mockClear()
        getModels.mockClear()

        done()
      })
    })
  })

  describe.each`
    value       | key        | wasMakeValid | fieldDataMake                                        | fieldDataModel
    ${null}     | ${'make'}  | ${false}     | ${{ values: makesList, valid: false, value: '' }}    | ${null}
    ${'f'}      | ${'make'}  | ${false}     | ${{ values: makesList, valid: false, value: 'f' }}   | ${null}
    ${'F'}      | ${'make'}  | ${false}     | ${{ values: makesList, valid: false, value: 'f' }}   | ${null}
    ${'Ford'}   | ${'make'}  | ${false}     | ${{ values: makesList, valid: true, value: 'ford' }} | ${{ values: modelsList }}
    ${'F'}      | ${'make'}  | ${true}      | ${{ values: makesList, valid: false, value: 'f' }}   | ${{}}
    ${'F'}      | ${'model'} | ${true}      | ${{ values: makesList, valid: true, value: 'ford' }} | ${{ values: modelsList, valid: false, value: 'f' }}
    ${'Fiesta'} | ${'model'} | ${true}      | ${{ values: makesList, valid: true, value: 'ford' }} | ${{ values: modelsList, valid: true, value: 'fiesta' }}
  `(
    `make.valid was $wasMakeValid, and handleUpdate receives value: $value, name: $key`,
    ({ value, key, wasMakeValid, fieldDataMake, fieldDataModel }) => {
      it(`updates state`, async done => {
        setGetMakesMock(makesList)
        setGetModelsMock(modelsList)

        const wrapper = await shallow(<TestedComponent {...getProps({})} />)
        wasMakeValid
          ? wrapper.instance().handleUpdate('Ford', 'make')
          : wrapper.instance().handleUpdate(value, key)

        process.nextTick(() => {
          wasMakeValid && wrapper.instance().handleUpdate(value, key)

          expect(wrapper.state()).toEqual({
            error: null,
            formData: {
              make: fieldDataMake,
              ...(fieldDataModel && { model: fieldDataModel }),
            },
            isLoading: false,
          })

          getMakes.mockClear()
          getModels.mockClear()

          done()
        })
      })
    }
  )
})
