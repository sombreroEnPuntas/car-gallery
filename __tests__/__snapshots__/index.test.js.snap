// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index matches snapshot 1`] = `
<Form
  isLoading={false}
  message={null}
>
  <p>
    Available makes:
  </p>
  <DropdownField
    autocomplete="make"
    disabled={false}
    handleUpdate={[Function]}
    name="make"
    valid={false}
    value=""
    values={
      Array [
        "ford",
        "opel",
      ]
    }
  />
  <p>
    Available models:
  </p>
  <DropdownField
    autocomplete="model"
    disabled={true}
    handleUpdate={[Function]}
    name="model"
    placeholder="Chose make first!"
    valid={false}
    value=""
    values={Array []}
  />
</Form>
`;
